bayes_kit/ensemble.py

```diff
@@ -1,68 +1,91 @@
     5(1):65--80.
     """
 
-    # def __init__(
-    #         self,
-    #         model: LogDensityModel,
-    #         a: Optional[float] = None,
-    #         walkers: Optional[int] = None
-    #         init: Optional[NDarray[np.float64]] = None)
-    #     ):
-    #     self._model = model
-    #     self._dim = self._model.dims()
-    #     if a != None and a < 1:
-    #         raise ValueError(f"stretch bound must be greater than or equal to 1; found {a=}")
-    #     self._a = a
-    #     self._sqrt_a = np.sqrt(a)
-    #     self._inv_sqrt_a = 1 / self._sqrt_a
-    #     if walkers != NONE and (walkers <= 0 or walkers % 2 != 0) :
-    #         raise ValueError(f"walkers must be strictly positive, even integer; found {walkers=}")
-    #     self._walkers = walkers or 2 * self._dim
-    #     self._halfwalkers = a / 2
-    #     self._drawshape = (self._walkers, self._dim)
-    #     if init != None and init.shape != self._drawshape:
-    #         raise ValueError(f"init must be shape of draw {self._drawshape}; found {init.shape=}")
-    #     self._thetas = init or np.random.normal(size=self._drawshape)
-    #     self._firsthalf = range(halfwalkers)
-    #     self._secondhalf = range(halfwalkers, walkers)
+    def __init__(
+            self,
+            model: LogDensityModel,
+            a: Optional[float] = None,
+            walkers: Optional[int] = None,
+            init: Optional[NDArray[np.float64]] = None
+        ):
+        """
+        Initialize the sampler with a log density model, and optionally
+        proposal bounds, number of walkers and initial parameter values.
+            
+        Parameters:
+        model: class used to evaluate log densities
+        a: bounds on proposal (default 2)
+        walkers: an even number of walkers to use (default dimensionality of `model * 2`)
+        init: `walker` x `dimensio`n array of initial positions (defaults to standard normal)
+
+        Throws:
+        ValueError: if `a` is provided and not >= 1, `walker`s is provided and not strictly positive and even,
+        or if the `init` is provided and is not an `NDArray` of shape `walker` x `dimension`
+        """
+        self._model = model
+        self._dim = self._model.dims()
+        if a != None and np.float64(a) < 1:
+            raise ValueError(f"stretch bound must be greater than or equal to 1; found {a=}")
+        self._a = np.float64(a or 2.0)
+        self._sqrt_a = np.sqrt(np.float64(a))
+        self._inv_sqrt_a = 1 / self._sqrt_a
+        self._walkers = np.int64(walkers or 2 * self._dim)
+        if self._walkers < 2 or self._walkers % 2 != 0:
+            raise ValueError(f"walkers must be strictly positive, even integer; found {walkers=}")
+        self._halfwalkers = self._walkers // 2
+        self._drawshape = (int(self._walkers), self._dim)
+        self._thetas = np.asarray(init or np.random.normal(size=self._drawshape))
```

Comment:
Would you mind doing something similar to what I did in #8 for these `np.random` calls (here and in the methods)? You should be able to basically copy some of the stuff done there.

Reply:
Sure.  I created this PR at the same time as #8 so it didn't have them initially.

Reply:
OK, replaced the RNG with seeded generator (constructor typing there is super flexible, which is convenient).